docs/architecture/state.md

```diff
@@ -1,91 +1,111 @@
 
 ![Architecture core concepts](../img/architecture/state/note-db.png)
 
-As described in [the notes section](notes.md), there are two types of notes:
+### Nullifier database
 
-* Public notes where the entire note content is recorded in the state.
-* Private notes where only a note's hash is recorded in the state.
+Each [note](notes.md) has an associated nullifier which enables the tracking of whether it's associated note has been consumed or not, preventing double-spending.
 
-As with accounts, there is a strong incentive to use private notes as they result in lower fees. This is also beneficial to the network as a private note adds only $64$ bytes to the state ($32$ bytes when it is produced, and $32$ bytes when it is consumed).
+To prove that a note has not been consumed, the operator must provide a Merkle path to the corresponding node and show that the node’s value is `0`. Since nullifiers are $32$ bytes each, the Sparse Merkle Tree height must be sufficient to represent all possible nullifiers. Operators must maintain the entire nullifier set to compute the new tree root after inserting new nullifiers.
 
-Using a Merkle Mountain Range (append-only accumulator) is important for two reasons:
+> **Note**
+> - Nullifiers in Miden break linkability between privately stored notes and their consumption details. To know the [note’s nullifier](notes.md#note-nullifier-ensuring-private-consumption), one must know the note’s data.
 
-1. Membership witnesses (that a note exists in the database) against such an accumulator needs to be updated very infrequently.
-2. Old membership witnesses can be extended to a new accumulator value, but this extension does not need to be done by the original witness holder.
+![Architecture core concepts](../img/architecture/state/nullifier-db.png)
 
-Both of these properties are needed for supporting local transactions using client-side proofs and privacy. In an append-only data structure, witness data does not become stale when the data structure is updated. That means users can generate valid proofs even if they don’t have the latest state of this database; so there is no need to query the operator on a constantly changing state.
+## Additional information
 
-However, the size of the note database does not grow indefinitely. Theoretically, at high tps, it grows very quickly: at $1$K TPS there are about $1$TB/year added to the database. But, only the unconsumed public notes, and enough info to construct membership proofs against them, need to be stored explicitly. Private notes, as well as public notes which have already been consumed, can be safely discarded. Such notes would still remain in the accumulator, but there is no need to store them explicitly as the append-only accumulator can be updated without knowing about all the items stored in it. This reduces actual storage requirements to a fraction of the database's nominal size.
+### Public shared state
```

Comment:
I like the idea of this section but I think it should be a part of a bigger section the describes how different users work with the state. Specifically, we could come up with different types of state:

1. Self-custodied state - state managed by a single user.
2. Federated state - state managed by a small number of users (e.g., under 100).
3. Network-managed state - state managed by the network itself (this would be the "public shared state").
4. Hybrids between the above (this could be a really cool thing).

Would be good to describe how these work and what are the tradeoffs between them.